crates/build/src/docker.rs

```diff
@@ -0,0 +1,346 @@
+use std::{
+    collections::{
+        hash_map::DefaultHasher,
+        HashMap,
+    },
+    hash::{
+        Hash,
+        Hasher,
+    },
+    path::Path,
+};
+
+use anyhow::{
+    Context,
+    Result,
+};
+use bollard::{
+    container::{
+        Config,
+        CreateContainerOptions,
+        ListContainersOptions,
+        StopContainerOptions,
+        WaitContainerOptions,
+    },
+    image::ListImagesOptions,
+    service::{
+        HostConfig,
+        Mount,
+        MountTypeEnum,
+    },
+    Docker,
+};
+use contract_metadata::ContractMetadata;
+use tokio_stream::StreamExt;
+
+use crate::{
+    maybe_println,
+    BuildArtifacts,
+    BuildResult,
+    BuildSteps,
+    CrateMetadata,
+    ExecuteArgs,
+    Network,
+    OptimizationPasses,
+    DEFAULT_MAX_MEMORY_PAGES,
+};
+
+use colored::Colorize;
+
+#[derive(Clone, Debug, Default)]
+pub enum ImageVariant {
+    #[default]
+    Default,
+    Custom(String),
+}
+
+/// Launched the docker container to execute verifiable build
+pub fn docker_build(args: ExecuteArgs) -> Result<BuildResult> {
+    let ExecuteArgs {
+        manifest_path,
+        verbosity,
+        features,
+        network,
+        unstable_flags,
+        optimization_passes,
+        keep_debug_symbols,
+        output_type,
+        target,
+        max_memory_pages,
+        build_artifact,
+        ..
+    } = args;
+    tokio::runtime::Runtime::new()?.block_on(async {
+        let mut build_steps = BuildSteps::new();
+
+        build_steps.set_total_steps(5);
+        if build_artifact == BuildArtifacts::CodeOnly {
+            build_steps.set_total_steps(4);
+        }
+
+        maybe_println!(
+            verbosity,
+            " {} {}",
+            format!("{build_steps}").bold(),
+            "Executing verifiable build".bright_green().bold()
+        );
+
+        let client = Docker::connect_with_socket_defaults()?;
+
+        // TODO: replace with image pulling, once the image is pushed to registry
+        let images = client
+            .list_images(Some(ListImagesOptions::<String> {
+                all: true,
+                ..Default::default()
+            }))
+            .await?;
+        let build_image = images.iter().find(|i| {
+            i.labels.get("io.parity.image.title")
+                == Some(&"contracts-verifiable".to_string())
+        });
+        if build_image.is_none() {
+            return Err(anyhow::anyhow!("No image found"))
+        }
+        let build_image = build_image.unwrap();
+
+        let crate_metadata = CrateMetadata::collect(&manifest_path, target)?;
+        let host_folder = crate_metadata
+            .manifest_path
+            .absolute_directory()?
+            .as_path()
+            .to_owned();
+
+        let file_path = host_folder.join(Path::new("target/build_result.json"));
```

Comment:
Did you make sure that user will be able to remove this file? ie that it's not created for user with incorrect privileges.

Reply:
https://doc.rust-lang.org/std/fs/struct.Permissions.html all files are created with o+r+w permissions, for the user that executes the command, so I don't think we'll have a problem there